Design.md

```diff
@@ -309,3 +335,36 @@ Get the platform usage of global (without query parameter), tenant (with tenant_
         * `key` : `offset` and `value` : `marker`
     1. `AssumeRoleCache` for all the `User` and `S3 Credential` APIs
         * `key` : `RoleArn` and `value` : `Credentials`
+
+## SecretKey Encryption Strategy
+
+1. `secret-key` value in the form of plaintext will be encrypted and decrypted using the cipher algorithm (property: `cipher`), and corresponding cryptographic key (property: `secretKey`) of the latest key slot in the `osis.security.keys` list provided in the `crypto.yml` config file.
+1. A `id` property will be provided for each key slot entry in the `osis.security.keys` list in the `crypto.yml` config file to maintain the version of the key and will be used in the Key rotation.
+    1. Example `crypto.yml`
+        ```yaml
+        keys:
+          - id: 2
+            cipher: AES256GCM
+            secretKey: YW5vdGhlcmxpbmVvZnBhc3N3b3JkZm9yYW5vdG==
+          - id: 1
+            cipher: AES256GCM
+            secretKey: dGhpc2lzYXJlYWxseWxvbmdhbmRzdHJvbmdrZXk=
+       ```
+1. A java object of class `SecretKeyRepoData` will be created with variables `keyId`, `encryptedData` and `cipherInformation`
+    * `keyId` variable will be of type String and stores the `id` property of the key used.
+    * `encryptedData` will be a byte array and stores the encrypted data bytes.
+    * `cipherInformation` will be an encapsulation to store all the additional information used by the cipher algorithm.
+        * Example: For `AES256GCM` algorithm, `cipherInformation` encapsulates the `nonce` variable, which will be a byte array and stores the nonce bytes.
+1. Java object of class `SecretKeyRepoData` will be serialized and stored as binary in Redis
+1. During decryption, `keyId` in the meta information will be used to identify applied the cipher algorithm.
+1. Initially, only the AES256GCM cipher algorithm is supported. (The Secretbox algorithm is suggested for future releases)
+1. An ability to change the cipher algorithm will be provided.
+1. **Key Rotation:**
+    * Key rotation is used to replace a key with another one once the original key is compromised or too old to use.
+    * `crypto.yml` file will be updated with a new key slot on top of `keys` list with an incremented `id` value.
+    * A key rotation application jar file will be provided to do the key rotation of all the encrypted `secret-key` data on Redis database. 
+        1. Key rotation application will use the meta-information of the encrypted data to identify the `keyId` used.
+        1. Each `secret-key` will be decrypted using the old cipher algorithm (identified using the `keyId`).
+        1. `secret-key` value will be re-encrypted using the latest cipher algorithm and updated on Redis.
+        1. This key rotation application need to be re-run until it returns success which means all the `secret-key` values on Redis are re-encrypted. 
+    * During decryption, if a new `keyId` is found, `OSIS` application will reload the `crypto.yml` file
```

Comment:
Will key rotation job be a background job with something like supervisors?

Reply:
A manual intervention is needed in this key rotation. If this needs to be a background job needs an automatic key rotation mechanism to be implemented like AWS KMS. I think we can add that integration when we implement KMS with automatic key rotation.

Reply:
Indeed, manual invocation as part of the key rotation procedure (which includes 'generating' a new random key and putting it in the configuration file(s)) should be fine for now. This procedure is to be written/documented, but should contain of roughly these steps:
1. Generate a new random key, encoded using base64: `openssl rand -base64 $(( 256 / 8 ))`
2. Add this new key in the secrets configuration, as a new (and the 'used') slot
3. Trigger OSIS to start using this new crypto slot for any new key it stores (and hence, encrypts)
4. Launch the rotation utility which will retrieve and (if needed) re-crypt and store old keys, now using the latest one
5. Repeat the previous until (successful) completion if needed
6. Remove the old key slot from the secrets configuration
7. Optionally, trigger OSIS to reload the config and fully forget the old, now-unused key slot